File: android-native/app/src/main/java/com/example/kpkn/data/db/PerformanceRangeEntity.kt
package com.example.kpkn.data.db

import androidx.room.Entity
import androidx.room.PrimaryKey
import kotlinx.serialization.encodeToString
import kotlinx.serialization.Serializable

@Serializable
data class PerformanceRangeData(
    val contextKey: String,
    val ermMin: Double,
    val ermMax: Double,
    val ermRms: Double,
    val sampleCount: Int = 0,
    val lastUpdatedMs: Long = 0L,
    val consecutiveAbove: Int = 0,
    val consecutiveBelow: Int = 0,
)

@Entity(tableName = "performance_range")
data class PerformanceRangeEntity(
    @PrimaryKey val contextKey: String,
    val data: String,
)

fun PerformanceRangeData.toEntity() = PerformanceRangeEntity(
    contextKey = contextKey,
    data = dbJson.encodeToString(this),
)

fun PerformanceRangeEntity.toPerformanceRangeData(): PerformanceRangeData =
    dbJson.decodeFromString(data)

File: ios-native/KPKNFit/KPKNFit/Data/Db/PerformanceRangeEntity.swift
import Foundation

// MARK: - JSON Helpers

private let dbJson: JSONEncoder = {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    encoder.dateEncodingStrategy = .iso8601
    return encoder
}()

private let dbJsonDecoder: JSONDecoder = {
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .iso8601
    decoder.keyDecodingStrategy = .useDefaultKeys
    return decoder
}()

private func encodeModel<T: Encodable>(_ value: T) -> String {
    guard let data = try? dbJson.encode(value) else { return "{}" }
    return String(decoding: data, as: UTF8.self)
}

private func decodeModel<T: Decodable>(_ type: T.Type, from data: String) -> T? {
    try? dbJsonDecoder.decode(type, from: Data(data.utf8))
}

// MARK: - PerformanceRangeData

internal struct PerformanceRangeData: Codable {
    let contextKey: String
    let ermMin: Double
    let ermMax: Double
    let ermRms: Double
    let sampleCount: Int
    let lastUpdatedMs: Int64
    let consecutiveAbove: Int
    let consecutiveBelow: Int
}

// MARK: - PerformanceRangeEntity

internal struct PerformanceRangeEntity: Codable {
    let contextKey: String
    let data: String
}

// MARK: - Conversions

extension PerformanceRangeData {
    func toEntity() -> PerformanceRangeEntity {
        PerformanceRangeEntity(contextKey: contextKey, data: encodeModel(self))
    }
}

extension PerformanceRangeEntity {
    func toPerformanceRangeData() -> PerformanceRangeData {
        decodeModel(PerformanceRangeData.self, from: data)!
    }
}
